PLANET-4919: Ensure automatically selected options are correctly shown in the Preview when switching themes

pablocubico opened the pull request:
Ref: https://jira.greenpeace.org/browse/PLANET-4919

This check was not working for Palette options, because the value is stored as `color`, and not as `value` in the `options` array (check the comment a few lines above):
https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/blob/develop/assets/src/components/PostMeta/withPostMeta.js#L56

Master theme PR: https://github.com/greenpeace/planet4-master-theme/pull/1053

Inwerpsel commented on assets/src/components/ColorPaletteControl/ColorPaletteControl.js:
```diff
@@ -7,13 +7,15 @@ import { ColorPalette } from '@wordpress/components';
 function ColorPaletteControl( { label, className, value, help, instanceId, onChange, options = [], ...passThroughProps } ) {
   const id = `inspector-color-palette-control-${ instanceId }`;
 
+  const optionsAsColors = options.map( option => ( { color: option.value } ) );
```
The colors passed to `ColorPalette` can have a label. I hadn't added that to the config files, as I didn't have the names for the colors we use. We may want to use these labels at some point as they improve Accessibility, so maybe you can already pass on this label from the theme here?

pablocubico replied:
So I did this change, but now I'm starting to regret it, as we introduced another thing to maintain (checking that color names match the actual colors, which is quite impossible to test automatically) and this whole UI is far from accessible... also is quite weird to imagine a visually impaired person choosing color palettes (although it's possible). Also, I found that none of the colors in the campaign templates are present in `_colors.scss`, not even the GP greens 🤦 , so I'm making a script to actually collect the colors we are actually using in the whole site. Anyway, it's done, but let's try to keep the scope limited to what the tickets need.

Inwerpsel replied:
The accessibility concerns are more broad in the sense that any user can benefit from it: likely we will have campaign themes with slight variations of the same color. Including a textual name will make it a lot easier to communicate about them, compared to having only the visual appearance to go by.

Maybe we could also set the text to the hex code if none is present.

pablocubico replied:
I think we are overthinking this a bit, there is no standard for color names, so we'll have to start naming things as "Navy Blue", or "Lime Green". If we don't pass a name, the swatch actually gets the Hex value in the tooltip, so maybe we should just remove the labels, or I could add the Hex color to the label, as in: `Red (#FF0000)`. 🤔

comzeradd replied:
Keep in mind also, that there is some design work pending here too, to vet our full color palette and keep the ones that we actually want to use. So I wouldn't spend much time trying to figure out names for colors that maybe shouldn't even be used.

Inwerpsel replied:
Indeed we should not be setting up the names right now, I was just suggesting to pass on the label already so that adding the labels at a later point becomes an easy configuration only task.

Yeah also noticed that it already shows the hex code. We can re-iterate later if we want it to display differently.

pablocubico commented:
Hey @Inwerpsel I added this line to the current commit, I think that's all it takes to update the meta on fields that depend on another: https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/249/files#diff-d9bfee53567c8981876ce5b3aade0e38R60

@comzeradd watch out: this is not approved yet but Github says it's ok to merge 🤔 

![image](https://user-images.githubusercontent.com/340766/79000401-ed62c880-7b22-11ea-95c9-27afde14ca32.png)

Inwerpsel commented on assets/src/components/ColorPaletteControl/ColorPaletteControl.js:
```diff
@@ -7,13 +7,15 @@ import { ColorPalette } from '@wordpress/components';
 function ColorPaletteControl( { label, className, value, help, instanceId, onChange, options = [], ...passThroughProps } ) {
   const id = `inspector-color-palette-control-${ instanceId }`;
 
+  const optionsAsColors = options.map( option => ( { color: option.value, name: option.label } ) );
```
Maybe an easier solution here is to spread so that we don't need to re-specify the `name` key and are future proof for other fields.
```
const { value, ...otherFields  } = option;
return {
    color: value,
    ...otherFields,
}
```

pablocubico replied:
Resolving this as we decided to remove the labels from the JSON files.

Inwerpsel commented on assets/src/components/PostMeta/withPostMeta.js:
```diff
@@ -57,6 +57,7 @@ export function withPostMeta( WrappedComponent ) {
         )
       ) {
         value = this.props.defaultValue;
+        this.handleChange( metaKey, value );
```
This definitely should not happen in `render`, as it will result in another render, and in general `render` should not have any side effects, even if they would not trigger a render.

Did you do this to trigger `this.props.writeMeta`or the preview update, or both?

pablocubico replied:
Resolving this as it changed and it's outdated now.

pablocubico commented:
Hey @Inwerpsel I did some changes in the last commit, mainly:
- removed the call to `handleThemeChange` in the `ThemeSelect` compo, because the themes were being loaded twice, because the `withPostMeta` fields are dispatching `editPost` actions to the store (via `writeMeta`), so the changes are catched by the `wp.data` subscriber on `CampaignSidebar`.
- in the `withPostMeta` HOC, I cached the value of the components in the HOC's state, using `getDerivedStateFromProps` which happens before render, and now I'm calling the `handleChange` method using `componentDidUpdate`, after render, which is a typical combination for handling side effects (lately replaced by the `useEffect` hook. 

I did a few tests and it looks like the preview is catching the right values now, while the saved post stays untouched. Let me know how it looks!

pablocubico commented on assets/src/components/PostMeta/withPostMeta.js:
```diff
@@ -14,13 +14,38 @@ const getValuePropName = ( Component ) => {
       this.handleChange = this.handleChange.bind( this );
       this.valuePropName = getValuePropName( WrappedComponent );
+      this.state = {
+        value: getValuePropName( props )
```
You're right, the call is wrong, and I think this is actually being overwritten by `getDerivedStateFromProps`. I'll test.

pablocubico replied:
Resolving as this is outdated.

pablocubico commented on assets/src/components/PostMeta/withPostMeta.js:
```diff
@@ -14,13 +14,38 @@ const getValuePropName = ( Component ) => {
       this.handleChange = this.handleChange.bind( this );
       this.valuePropName = getValuePropName( WrappedComponent );
+      this.state = {
+        value: getValueFromProps( props )
```
Hey @Inwerpsel I think your previous comment here got lost, I updated this to `getValueFromProps` but I'm keeping this in the constructor since otherwise it would trigger a warning: `react-dom.js:539 Warning: `WrappingComponent` uses `getDerivedStateFromProps` but its initial state is undefined. This is not recommended. Instead, define the initial state by assigning an object to `this.state` in the constructor of `WrappingComponent`. This ensures that `getDerivedStateFromProps` arguments have a consistent shape.`

Inwerpsel replied:
You could still assign `null` to `value` so that `getValueFromProps` isn't executed twice?

pablocubico replied:
This got outdated too.

Inwerpsel commented on assets/src/components/PostMeta/withPostMeta.js:
```diff
@@ -14,13 +14,38 @@ const getValuePropName = ( Component ) => {
+      };
     }
 
     handleChange( metaKey, value ) {
```
We can actually adapt the code below that updates the preview meta to use the function I introduced to do the same thing on page load: https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/blob/0c211c994b7626a37b5c8b3f886181e24791a84a/assets/src/saveMetaToPreview.js

But maybe it actually would make more sense if we set this up as a listener to the `wp/data` store in some way? Then we wouldn't even need to call `savePreviewMeta` more than once. Though a challenge with that approach is that you need to listen to any event in the `wp/data` store and then only fire if the meta has changed. Or maybe there is a way after all to just listen post meta changes.

Inwerpsel replied:
It would also be an improvement over the current call to `savePreviewMeta` happening on the `DOMContentLoaded` event.

pablocubico commented:
Hey @Inwerpsel news about this feature:

- I removed the `DOMContentLoaded` subscriber from the `savePreviewMeta` method
- Removed the duplicated coda (the one POSTing to `/save-preview-meta`
- Added the  `savePreviewMeta` here: https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/249/files#diff-0fe7b281274ce46f740cd4828efc6356R102 , where we are setting the `meta` for the CampaignSidebar.

The call to `getDerivedStateFromProps` is still there as I'm still relying on the cached state value to call `handleChange` which writes the meta. But so far it looks good, maybe we can wrap it up if all these works, I'm currently testing it.

Inwerpsel commented on assets/src/components/PostMeta/withPostMeta.js:
```diff
@@ -14,56 +14,66 @@ const getValuePropName = ( Component ) => {
-        value = this.props.defaultValue;
-      } else {
-        value = metaValue;
+      if (metaValue != this.state.value) {
```
Can you make this a strict comparison? `!==`

Inwerpsel commented on assets/src/components/Sidebar/CampaignSidebar.js:
```diff
@@ -85,19 +86,20 @@ export class CampaignSidebar extends Component {
         }
         if ( !isShallowEqual( this.state.meta, meta ) ) {
           this.setState( { meta: meta } );
+          savePreviewMeta();
```
Be aware that `CampaignSidebar` will only mount if the icon is clicked! I think we might also need to update this preview if the sidebar isn't opened.

pablocubico replied:
Hmmm good point, so maybe we need to call it on `DOMContentLoaded` after all. Or maybe there is a better editor hook for that. I'll dig into it.

pablocubico replied:
I just checked: looks like the component is mounted even if the Sidebar is not displayed, so `savePreviewMeta` is called on load.

pablocubico replied:
This is not relevant anymore.

pablocubico commented:
Hey @Inwerpsel , I merged your branch, squashed some of the commits and rebased to master.

Inwerpsel commented on assets/src/components/PostMeta/withPostMeta.js:
```diff
@@ -21,51 +33,35 @@ export function withPostMeta( WrappedComponent ) {
       this.handleChange = this.handleChange.bind( this );
       this.valuePropName = getValuePropName( WrappedComponent );
+      this.state = {
+        value: null
```
Still not sure if we need to introduce state and `getDerivedStateFromProps` in the component to do handle translating the props into a value. What is the benefit over calling `getValueFromProps` in render?

pablocubico replied:
This is not needed anymore as you removed the `componentDidUpdate` method in your previous commit.
